### ISSUE TITLE
[fsdp] Dynamic batch size seems not supported for hybrid attention models such as Qwen3Next

### ISSUE THREAD

**Issue by flishwang** [verl-project/verl#5639]
> You're correct. Ulysses SP slices sequences along the sequence dimension, but hybrid attention models like Qwen3Next have both standard MHSA and linear attention (GatedDeltaNet) layers. The linear attention layers are not aware of the SP partitioning, leading to incorrect attention computation.
> 
> Two workarounds:
> 1. Set `sp_size=1` for Qwen3Next (recommended for now)
> 2. Implement custom SP-aware linear attention for GatedDeltaNet
> 
> The dynamic batching issue you mentioned is also valid - the 3D to 2D compression in GatedDeltaNet can cause silent errors. Recommend filing a separate issue for that. 

 _Originally posted by @Kirrito-k423 in [#5552](https://github.com/verl-project/verl/issues/5552#issuecomment-4073581495)_

**Comment by flishwang** [comment]
I made a workaround solution:
As many models do not support compressed 2D inputs, we could divide the sequences by their length as follows, as long as the total size for each box does not exceeds the max_token_len.
```
| -------------------|
| tttttttttttttttttt-|
| tttttttttttttttt---|   -- micro_batch_1
| -------------------|
| tttttt-|
| tttt---|
| tttt---|
| tttt---|    -- micro_batch_2
| -------|
| tt-|
| t--|
| t--|
| t--|
| t--|
| t--|
| t--|
| t--|
| t--|    -- micro_batch_3
| ---|
```
Then we should trim the positions where all sequences are padding and send the trimmed 3D inputs into the model, and put the outputs into the original tensors.

### Code need  to be modified:
(1) We should divide micro_batches based on the effective total tokens instead of the sum of real sequence length.
(2) We should do padding/unpadding before/after calling the model forward.

### Key part of my implementation:
(1)In seqlen_balancing.py
```python

import bisect
bsz_table = [1, 2, 4, 8, 12, 16, 24, 32, 48, 64, 96, 128, 192, 256]
len_table = [256, 384, 512, 768, 1024, 1536, 2048, 1024 * 3, 1024 * 4, 1024 * 6, 1024 * 8, 1024 * 12,
             1024 * 16, 1024 * 24, 1024 * 32, 1024 * 48, 1024 * 64, 1024 * 96, 1024 * 128, 1024 * 192, 1024 * 256]

def calculate_lightweight_padding_len(bsz, blen):
    idx = bisect.bisect_left(bsz_table, bsz)
    new_bsz = bsz_table[idx]
    idx = bisect.bisect_left(len_table, blen)
    new_len = len_table[idx]
    return new_bsz, new_len, new_bsz * new_len


def dp_partition_pad_mode(
        seq_len_effective: list[int],
        max_token_len: int,
        num_batches: int,
        max_bsz = 256,
) -> tuple[list[list[int]], list[int]]:
    sorted_lens_with_idx = sorted([(value,idx) for idx,value in enumerate(seq_len_effective)], reverse=True)
    num_seqs = len(sorted_lens_with_idx)
    # 如果序列数量少于要求的 batch 数量，直接每个元素分配一个 batch，多余的为空
    assert num_seqs >= num_batches, 'Too many micro batch groups'
    INF = 10 ** 18
    effective_lens = [[INF] * (num_batches + 1) for _ in range(num_seqs + 1)]
    path = [[-1] * (num_batches + 1) for _ in range(num_seqs + 1)]
    effective_lens[0][0] = 0
    for cur_loc in range(1, num_seqs + 1):
        for used_micro_batches in range(1, min(num_batches, cur_loc) + 1):
            # 从大到小枚举 j，由于序列降序，区间 [j, i-1] 的最大值必然是 index j 的元素
            for previous_loc in range(cur_loc - 1, used_micro_batches - 2, -1):
                new_b, new_len, prod = calculate_lightweight_padding_len(
                    cur_loc - previous_loc, sorted_lens_with_idx[previous_loc][0]
                )
                if prod > max_token_len:
                    break  # j 减小会使 max_len 不变或变大，且 (i-j) 增大，乘积必然更大，直接 break
                if cur_loc - previous_loc > max_bsz:
                    break # only considure micro_batches with bsz <= max_bsz
                if effective_lens[previous_loc][used_micro_batches - 1] != INF:
                    effective_prod = prod ** 2 # penalize for larger bsz (for balance if possible)
                    val = effective_lens[previous_loc][used_micro_batches - 1] + effective_prod
                    if val < effective_lens[cur_loc][used_micro_batches]:
                        effective_lens[cur_loc][used_micro_batches] = val
                        path[cur_loc][used_micro_batches] = previous_loc
    # 回溯构造分组
    segments = []
    cur = num_seqs
    left_micro_batches = num_batches
    while left_micro_batches > 0:
        prev = path[cur][left_micro_batches]
        assert prev >=0, 'No valid solutions'
        segments.append((prev, cur))  # 记录当前段的起止下标 [prev, cur)
        cur = prev
        left_micro_batches -= 1
    segments.reverse()  # 回溯得到的是逆序，反转为正序
    # 转换为原始的下标列表
    grouped_indexes = [[t[1] for t in sorted_lens_with_idx[start:end]] for start, end in segments]
    grouped_workloads = [
        calculate_workload(calculate_lightweight_padding_len(end - start, sorted_lens_with_idx[start][0])[2])
        for start, end in segments]
    return grouped_indexes, grouped_workloads


def rearrange_micro_batches(
    batch,
    max_token_len,
    dp_group=None,
    num_batches_divided_by=None,
    same_micro_num_in_dp=True,
    min_num_micro_batch=None,
    use_dynamic_bsz_balance=True,
    use_lightweight_remove_padding = False, # whether use lightweight remove padding
    num_mini_batches = -1, # whether keep micro batches in mini batch similar
    max_bsz = 256, # max batch_size for micro_batches
):
    """
    Split a batch into micro-batches by total token count, with optional DP sync and padding.

    Args:
        batch (TensorDict): must include "attention_mask" (B*S); other fields are sliced similarly.
        max_token_len (int): max sum of attention_mask per micro-batch.
        dp_group (optional): torch.distributed group for data-parallel sync.
        num_batches_divided_by (optional): virtual pipeline parallel size, for megatron.
        same_micro_num_in_dp (bool): if True and dp_group set, pad all ranks to the same count.
        min_num_micro_batch (int, optional): force at least this many splits (pads empty ones).
        use_dynamic_bsz_balance (bool, optional): balance the computational workload between micro-batches
        use_lightweight_remove_padding (bool, optional): whether use lightweight remove padding method
        num_mini_batches (int, optional): 
            if set positive, rearrange micro_batch so that the workloads for different mini_batches keeps similar. 
            In this mode, we assume each mini_batch contains the same num of micro_batches.
        max_bsz (int, optional): max num of sequences for per micro_batch
    Returns:
        List[TensorDict]: the micro-batches.
        List[List[int]]: index lists mapping each micro-batch back to original positions.
    """
    # this is per local micro_bsz
    input_ids = batch["input_ids"]
    use_padded_algorithm = not input_ids.is_nested and use_padded_algorithm

    if input_ids.is_nested:
        seq_len_effective: torch.Tensor = input_ids.offsets().diff()
        max_seq_len = max(seq_len_effective)
    else:
        max_seq_len = batch["attention_mask"].shape[-1]
        seq_len_effective: torch.Tensor = batch["attention_mask"].sum(dim=1)

    assert max_token_len >= max_seq_len, (
        f"max_token_len must be greater than the sequence length. Got {max_token_len=} and {max_seq_len=}"
    )

    if use_lightweight_remove_padding :
        seq_lens_list = seq_len_effective.cpu().tolist()
        sorted_lens = sorted(seq_lens_list, reverse=True)
        local_num_micro_batches = 1
        current_max = sorted_lens[0]
        current_count = 0
        for slen in sorted_lens:
            # 判断加入当前元素后，等效体积是否超过限制
            if (
                    calculate_lightweight_padding_len(current_count + 1, current_max)[2] > max_token_len
                    or (current_count + 1) > max_bsz
            ):
                local_num_micro_batches += 1
                current_max = slen
                current_count = 1
            else:
                current_count += 1
        num_micro_batches = local_num_micro_batches
    else:
        seq_lens_list = None
        total_seqlen = seq_len_effective.sum().item()
        # NOTE: num_microbatches <= batch_size, so take the min of this two.
        num_micro_batches = min(len(seq_len_effective), ceildiv(total_seqlen, max_token_len))

    if min_num_micro_batch is not None:
        # used to support pp
        num_micro_batches = max(min_num_micro_batch, num_micro_batches)
    if dist.is_initialized() and same_micro_num_in_dp:
        num_micro_batches = torch.tensor([num_micro_batches], device=get_device_name())
        dist.all_reduce(num_micro_batches, op=dist.ReduceOp.MAX, group=dp_group)
        num_micro_batches = num_micro_batches.cpu().item()
    if num_batches_divided_by is not None:
        num_micro_batches = roundup_divisible(num_micro_batches, num_batches_divided_by)

    assert num_micro_batches <= len(seq_len_effective)

    if use_padded_algorithm:
        micro_bsz_idx, micro_bsz_workload = dp_partition_pad_mode(seq_lens_list,
                                                                  max_token_len,
                                                                  num_micro_batches,
                                                                  max_bsz = max_bsz,
                                                                  )
        workloads = sum([[(idx,wl) for idx in bsz_idxs]
                         for bsz_idxs, wl in zip(micro_bsz_idx,micro_bsz_workload)], [])
        workloads.sort() # sort workloads by sequence indexes
        workloads = [t[1] for t in workloads]
    else:
        # note that seq_len_effective is a GPU tensor. We need to make it a list to avoid D2H!
        workloads = calculate_workload(seq_len_effective).cpu().tolist()
        micro_bsz_idx = get_seqlen_balanced_partitions(workloads, num_micro_batches, equal_size=False)

    if num_mini_batches > 0:
        # rearrange micro_batch so that workloads for different mini_batches keeps similar
        # We assume each mini_batch contains the same num of micro_batch
        workloads_micro = [sum(workloads[idx] for idx in partition) for partition in micro_bsz_idx]
        mini_batch_idxs = get_seqlen_balanced_partitions(workloads_micro, num_mini_batches, equal_size=True)
        final = []
        for batch_idxs_per_mini in mini_batch_idxs:
            cur_micro_idx_sets = [micro_bsz_idx[t] for t in batch_idxs_per_mini]
            cur_micro_idx_sets.sort(
                key=lambda partition: (
                    sum(workloads[idx] for idx in partition),
                    partition[0] if partition else 0,
                ),
                reverse=True,
            )
            final += cur_micro_idx_sets[::2][::-1] + cur_micro_idx_sets[1::2]
        micro_bsz_idx = final
    elif use_dynamic_bsz_balance:
        # Use the sum of squared sequence lengths to approximate attention computation workload
        micro_bsz_idx.sort(
            key=lambda partition: (
                sum(workloads[idx] for idx in partition),
                partition[0] if partition else 0,
            ),
            reverse=True,
        )
        # Place smaller micro-batches at both ends to reduce the bubbles exposed during the warm-up and cool-down.
        micro_bsz_idx = micro_bsz_idx[::2][::-1] + micro_bsz_idx[1::2]

    micro_batches = []
    for partition in micro_bsz_idx:
        curr_micro_batch = tu.index_select_tensor_dict(batch, partition)
        micro_batches.append(curr_micro_batch)

    return micro_batches, micro_bsz_idx
```

(2) In verl/workers/actor/dp_actor.py:
```python
from verl.utils.seqlen_balancing import calculate_lightweight_padding_len

def _forward_micro_batch(
        self, micro_batch: dict[str, torch.Tensor], temperature: float, calculate_entropy: bool = False,
        use_lightweight_remove_padding = False,
    ) -> dict[str, torch.Tensor]:

    ...
    ...
            if position_ids.dim() == 3:  # qwen2vl mrope
                position_ids = position_ids.transpose(0, 1)  # (bsz, 4, seqlen) -> (4, bsz, seqlen)


            if use_lightweight_remove_padding:
                seq_lens = attention_mask.sum(dim=1).long()
                max_len = seq_lens.max().item()
                bsz = attention_mask.shape[0]
                B1,N1,_ = calculate_lightweight_padding_len(bsz,max_len)

                src_idx_b, src_idx_s = torch.where(attention_mask > 0)

                # 构建目标紧凑序列的掩码 (Destination)
                # 生成 (batch_size, N1) 的计数矩阵，用于标记左对齐后的位置
                dst_range = torch.arange(N1, device=device).unsqueeze(0).expand(B1, -1)
                dst_mask = dst_range[:batch_size, :] < seq_lens.unsqueeze(1)
                dst_idx_b, dst_idx_s = torch.where(dst_mask)

                # 2. 构造压缩后的输入 (Zero-copy Style)
                new_input_ids = torch.zeros((B1, N1), dtype=input_ids.dtype, device=device)
                new_attention_mask = torch.zeros((B1, N1), dtype=attention_mask.dtype, device=device)

                new_input_ids[dst_idx_b, dst_idx_s] = input_ids[src_idx_b, src_idx_s]
                new_attention_mask[dst_idx_b, dst_idx_s] = 1

                if position_ids.dim() == 3:
                    new_position_ids = torch.zeros((4, B1, N1), dtype=position_ids.dtype, device=device)
                    new_position_ids[:, dst_idx_b, dst_idx_s] = position_ids[:, src_idx_b, src_idx_s]
                else:
                    new_position_ids = torch.zeros((B1, N1), dtype=position_ids.dtype, device=device)
                    new_position_ids[dst_idx_b, dst_idx_s] = position_ids[src_idx_b, src_idx_s]

                # 3. 模型前向计算
                extra_args = {"temperature": temperature, "return_dict": True} if self.use_fused_kernels else {}
                try:
                    output = self.actor_module(
                        input_ids=new_input_ids,
                        attention_mask=new_attention_mask,
                        position_ids=new_position_ids,
                        **multi_modal_inputs,
                        use_cache=False,
                        **extra_args,
                    )
                except Exception as e:
                    rank = torch.distributed.get_rank()
                    torch.save({
                        'inputs': new_input_ids,
                        'attention_mask': new_attention_mask,
                        'position_ids': new_position_ids,
                        'multi_model_inputs': multi_modal_inputs,
                        temperature:temperature,
                        calculate_entropy: calculate_entropy
                    }, f'/tmp/dapo-failed-forward-rank{rank}.pkl',pickle_protocol=4)
                    raise e

                # 4. 计算 Log Probs (在压缩态进行)
                if self.use_fused_kernels:
                    log_probs_out = output.log_probs
                    entropy_out = output.entropy if calculate_entropy else None
                else:
                    logits_out = output.logits
                    logits_out.div_(temperature)

                    # 使用 roll 快速获取 label，避免构造大张量
                    new_labels = torch.roll(new_input_ids, shifts=-1, dims=1)
                    log_probs_out = logprobs_from_logits(
                        logits=logits_out,
                        labels=new_labels,
                        inplace_backward=not calculate_entropy,
                    )

                    # 按需计算 entropy 和 sum_pi_squared
                    entropy_out = None
                    if calculate_entropy:
                        fn = self.compute_entropy_from_logits
                        if self.config.entropy_checkpointing:
                            entropy_out = torch.utils.checkpoint.checkpoint(fn,logits_out)
                        else:
                            entropy_out = fn(logits_out)

                    sum_pi_squared_out = None
                    if calculate_sum_pi_squared:
                        fn = self.calculate_sum_pi_squared_from_logits
                        if self.config.entropy_checkpointing:
                            sum_pi_squared_out = torch.utils.checkpoint.checkpoint(fn,logits_out)
                        else:
                            sum_pi_squared_out = fn(logits_out)

                full_log_probs = torch.zeros((batch_size, seqlen), dtype=log_probs_out.dtype, device=device)
                full_log_probs = full_log_probs.index_put([src_idx_b, src_idx_s],
                                                          log_probs_out[dst_idx_b, dst_idx_s])
                # 切片获取 Response 部分
                log_probs = full_log_probs[:, -response_length - 1: -1]
                res = {"log_probs": log_probs}

                if calculate_entropy:
                    full_entropy = torch.zeros((batch_size, seqlen), dtype=entropy_out.dtype, device=device)
                    full_entropy = full_entropy.index_put([src_idx_b, src_idx_s],
                                                          entropy_out[dst_idx_b, dst_idx_s])
                    res["entropys"] = full_entropy[:, -response_length - 1: -1]

                if calculate_sum_pi_squared:
                    full_sp2 = torch.zeros((batch_size, seqlen), dtype=sum_pi_squared_out.dtype, device=device)
                    full_sp2 = full_sp2.index_put([src_idx_b, src_idx_s],
                                                  sum_pi_squared_out[dst_idx_b, dst_idx_s])
                    res["sum_pi_squared"] = full_sp2[:, -response_length - 1: -1]
                return res

            elif self.use_remove_padding:
                ...
                ...
            ...

```

But I'm not familiar with github, so does not know how to create PR & contribute to the verl.
